src/ppx/grfx/grfx_foveation.cpp

```diff
@@ -0,0 +1,335 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ppx/grfx/grfx_foveation.h"
+#include "ppx/grfx/grfx_device.h"
+
+namespace ppx {
+namespace grfx {
+
+Result FoveationPattern::CreateApiObjects(const FoveationPatternCreateInfo* pCreateInfo)
+{
+    mFoveationMode = FOVEATION_NONE;
+    TexturePtr foveationTexture;
+
+    if (pCreateInfo->foveationMode == FOVEATION_FDM) {
+        Result ppxres = CreateFDMTexture(pCreateInfo, &foveationTexture);
+        if (Failed(ppxres)) {
+            PPX_LOG_INFO("Failed to create default foveation textue.");
+            return ppxres;
+        }
+    }
+    else if (pCreateInfo->foveationMode == FOVEATION_VRS) {
+        Result ppxres = CreateVRSTexture(pCreateInfo, &foveationTexture);
+        if (Failed(ppxres)) {
+            PPX_LOG_INFO("Failed to create default foveation textue.");
+            return ppxres;
+        }
+    }
+    mFoveationTexture = foveationTexture;
+    mFoveationMode    = pCreateInfo->foveationMode;
+
+    std::vector<std::uint8_t> data(mTextureWidth * mTextureHeight * mTexelSize);
+    FillFoveationTextureData(pCreateInfo->patternType, data.data());
+    Result ppxres = UploadFoveationTexture(data.data(), true);
+    if (Failed(ppxres)) {
+        PPX_ASSERT_MSG(false, "Foveation: texture update failed");
+    }
+
+    return ppx::SUCCESS;
+}
+
+void FoveationPattern::DestroyApiObjects()
+{
+    if (mFoveationTexture) {
+        GetDevice()->DestroyTexture(mFoveationTexture);
+        mFoveationTexture.Reset();
+    }
+}
+
+Result FoveationPattern::CreateFDMTexture(const FoveationPatternCreateInfo* pCreateInfo, Texture** ppFoveationTexture)
+{
+    const FoveationCapabilities& foveationCapabilites = GetDevice()->GetFoveationCapabilities();
+    if (!foveationCapabilites.fdm.supported) {
+        PPX_LOG_WARN("Density map not supported by GPU. Disable foveation.");
+        return ppx::SUCCESS;
+    }
+
+    // Create foveation texture
+    uint32_t texel_w = foveationCapabilites.fdm.texelSize.min.width;
+    uint32_t texel_h = foveationCapabilites.fdm.texelSize.min.height;
+    mTextureWidth    = pCreateInfo->fbWidth / texel_w;
```

Comment:
Should this be `ceil` instead of `floor`?

Reply:
I think floor is correct because `texel_w` is the minimum width in the render target that can correspond to a single pixel in the FDM, so `fbWidth / texel_w` is the maximum allowed width of the FDM, so ceil would make the FDM too wide.

Reply:
Discussed in chat, according to [vkspec] and [Quacomm extension], texture width should be `((fbWidth + texel_w - 1)/texel_w)`

[vkspec]: https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#fragmentdensitymap-fetch-density-value
[Quacomm extension]: https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap12.html#VUID-VkImageCreateInfo-fragmentDensityMapOffset-06514

Reply:
This now uses ceil (src/ppx/grfx/vk/vk_shading_rate.cpp:34).